### ISSUE TITLE
feat: local ML training pipeline — custom Hey Jandal wake word model

### ISSUE THREAD

**Issue by lokhor** [NickMonrad/kernel-ai-assistant#984]
## Summary

Set up the local ML training pipeline and produce `hey_jandal.onnx` — the custom wake word classifier that runs as the third stage of the openWakeWord inference pipeline. This is an ML/data workstream that runs in parallel with the Android code work (#985).

"Hey Jandal" cannot be sourced from any pre-trained model library — it is an invented word with NZ-specific phonetics (Māori-influenced short front /a/, distinct from both American and British pronunciation). The model must be trained from synthetic data generated locally.

## Platform

**Local machine:** AMD Ryzen 7 5700X3D, AMD Radeon RX 9070 XT (RDNA 4 / Navi 48), Linux.

**ROCm note:** Verify ROCm support for RDNA 4 first — `rocminfo` should list the GPU. If ROCm 6.3+ is available and stable, use it for training. If not, CPU training is the correct fallback — openWakeWord classifiers are tiny (<1MB) and the 5700X3D handles this without issue.

## Architecture — 3-stage ONNX pipeline

openWakeWord is **not** a single monolithic model. It is a fixed 3-stage pipeline:

```
Audio (16kHz, 80ms = 1280 samples per chunk)
  → [Stage 1] melspectrogram.onnx      — raw PCM → mel-spectrogram patch
  → [Stage 2] embedding_model.onnx     — mel patch → 96-dim embedding vector (1 per 80ms frame)
  → ring buffer (last 28 frames = ~2.24s of acoustic context)
  → [Stage 3] hey_jandal.onnx          — 28×96 embedding window → confidence [0, 1]
```

**You only train Stage 3.** Stages 1 and 2 are pre-built shared models provided by the openWakeWord project and must be downloaded separately. All three files must be present at runtime.

### Model files needed at runtime

| File | Source | Size |
|------|--------|------|
| `melspectrogram.onnx` | Download from [openWakeWord releases](https://github.com/dscripka/openWakeWord/releases) | ~170 KB |
| `embedding_model.onnx` | Download from [openWakeWord releases](https://github.com/dscripka/openWakeWord/releases) | ~35 MB |
| `hey_jandal.onnx` | **Produced by this pipeline** | <1 MB |

Place all three in `app/src/main/assets/models/wakeword/`.

### Model contracts (Stage 3 input/output)

The classifier (`hey_jandal.onnx`) receives **pre-computed embeddings**, not raw audio:
- **Input:** `float32[1, 28, 96]` — sliding window of 28 embedding frames from Stage 2
- **Output:** `float32[1, 1]` — confidence that the wake phrase was spoken, in [0, 1]

The backbone (Stage 2) processes 80ms frames at 16kHz (1280 samples/frame) and produces one 96-dim embedding per frame. The classifier sees the most recent 28 frames (~2.24s of context).

**Audio is 16kHz, mono, 16-bit PCM — mandatory and non-negotiable.** This is a hard requirement of the openWakeWord pipeline, not a preference.

### Option A — openWakeWord.com (recommended for first model)

[openWakeWord.com](https://openWakeWord.com) provides a web interface that handles synthetic voice generation (via Piper) and cloud-GPU training, returning a ready-to-use `.onnx` classifier. Fastest path to a working model. Limitations: less control over negative examples and false positive tuning.

### Option B — Google Colab notebook

The [official Colab notebook](https://github.com/dscripka/openWakeWord) walks through synthetic data generation and training step by step. Good for parameter iteration without local GPU setup.

### Option C — fully local (preferred for iteration + privacy)

```bash
# Python 3.10+
pip install openwakeword torch torchaudio onnx onnxruntime
git clone https://github.com/dscripka/openWakeWord
cd openWakeWord

# ROCm path (only if rocminfo confirms RDNA4 GPU):
pip install torch --index-url https://download.pytorch.org/whl/rocm6.3
```

Training is done via the project's Jupyter notebook (`notebooks/training_models.ipynb`), not a CLI script. Run it locally with Jupyter:

```bash
pip install jupyter
jupyter notebook notebooks/training_models.ipynb
```

Key notebook parameters:
- `target_phrase` — the wake phrase string
- `n_samples` — number of synthetic positive examples to generate (500–1000)
- `false_positive_weight` — increase to suppress false positives at the cost of sensitivity (start at 5.0, tune from real-device FP rate in #986)

## Step 1 — Positive example generation: "Hey Jandal"

Generate 500–1000 diverse audio samples. Voice diversity matters more than volume — a model trained on 3 voices will overfit.

**Phonetics to preserve:** "Jandal" = /ˈdʒændl/ — the short front /a/ as in "jam", not the lengthened /ɑː/ common in British or American speech. This is the critical differentiator for NZ recognition.

### TTS sources for NZ-sounding samples

- **Voicebox** ([jamiepine/voicebox](https://github.com/jamiepine/voicebox)) + **Qwen3TTS** — voice cloning from real kiwi speaker recordings; best path to authentic NZ phonetics
- **Sherpa-ONNX VCTK/Semaine voices** — already in project assets; NZ phonetic tuning from #784 feeds directly into sample quality
- **Piper NZ English model** — if available; lightweight and fast
- **Additional diversity:** Coqui XTTS, Kokoro — vary speaker, pitch, and rate

### Augmentation per sample

Convolve with room impulse responses (0.5m–5m simulated distance), mix background noise at −10dB to +5dB SNR, vary playback speed ±20%.

## Step 2 — Negative examples

openWakeWord uses ambient audio as negatives. Essential inclusions:
- `audioset` or `fma` clips — speech and non-speech
- **Similar-sounding phrases** (critical): "hey Daniel", "hey handle", "hey Sandra", "hey gentle", "hey Janet", "hey candle"
- NZ-accent general conversation — the model must not fire on normal NZ speech

## Step 3 — Training (notebook)

Run `notebooks/training_models.ipynb`. The notebook:
1. Calls `openwakeword.utils.get_pretrained_model_paths()` to locate the shared melspectrogram/embedding models (downloads if absent)
2. Computes embeddings for all positive and negative clips using the shared backbone
3. Trains a small FCN classifier (PyTorch) on the pre-computed embeddings
4. Exports the classifier to ONNX via `torch.onnx.export`

The exported file is the Stage 3 classifier only. The shared Stage 1/2 models are managed separately.

## Step 4 — Export and quantisation

The notebook handles the initial ONNX export. After export, optionally quantise:

```python
from onnxruntime.quantization import quantize_dynamic, QuantType
quantize_dynamic("hey_jandal.onnx", "hey_jandal_int8.onnx", weight_type=QuantType.QInt8)
```

Target: <1MB classifier, <5ms total pipeline inference per 80ms chunk on Adreno 740 via NNAPI.

> **Note:** The quantised `hey_jandal_int8.onnx` and unquantised `hey_jandal.onnx` are both valid. The Android detector (`OnnxWakeWordDetector`) loads whichever file is placed at `assets/models/wakeword/hey_jandal.onnx` — rename accordingly.

## Step 5 — Validation before bundling

| Metric | Test | Target |
|--------|------|--------|
| False positive rate | 10 minutes continuous NZ speech | <1 activation/hour |
| True positive rate | 50 manual recordings, primary user | >95% |
| Threshold | Record confidence score distribution on true positives | Document recommended runtime threshold |

FP rate iteration is tracked in #986 (false positive tuning). That issue is gated on real-device measurements from this model.

## Step 6 — Bundle

Place **all three files** in `app/src/main/assets/models/wakeword/`:
- `melspectrogram.onnx`
- `embedding_model.onnx`
- `hey_jandal.onnx`

Document the recommended confidence threshold. The runtime default is configurable in Settings (#985).

## Deliverable

- `hey_jandal.onnx` — validated classifier, bundled with shared backbone models in `app/src/main/assets/models/wakeword/`
- Recommended confidence threshold value documented in a PR comment or `WakeWordPreferences` default
- End-to-end smoke test: device wakes on "Hey Jandal" spoken naturally

## Related

- Parent: #65 — Hey Jandal / Android Assistant epic
- Blocks: end-to-end testing of #985 (code infrastructure exists; model needed for device testing)
- Feeds: #986 — false positive rate tuning (needs real-device FP measurements from this model)
- Related: #784 — Kiwi language corpus tuning (NZ TTS phonetic work feeds positive example quality)

**Comment by NickMonrad** [comment]
Some resources we should review to make sure we are clear on options. 

I also want to consider if there's a way to use qwen3tts to clone kiwi voices to generate additional kiwi voices samples for training the model: 
https://github.com/jamiepine/voicebox

Here are some of the most useful and authoritative resources for diving deeper into openWakeWord, model training, and integrating it all into Jandal AI and Home Assistant:
## openWakeWord & Model Training
 * **openWakeWord Official GitHub Repository**
       This is the core project by David Scripka. It is the best place to understand the underlying architecture, how the models are structured, and the technical requirements for running inference.
 * **OpenWakeWord.com**
   The most user-friendly way to train a custom model. It provides a simple web interface that handles the synthetic voice generation (via Piper) and cloud-GPU training, returning a ready-to-use .onnx file.
 * **Automated Model Training (Google Colab)**
       Maintained by the openWakeWord project, this Colab notebook walks you step-by-step through generating synthetic data and training the model using cloud resources. It is great if you want to tweak parameters like the false positive rate or add custom background noises.
 * **Training Models Notebook (Local Execution)**
   If you want to leverage your Ryzen 5700X3D and train entirely locally, this Jupyter Notebook from the official repository details how to manually mix datasets and execute the training pipeline on your own hardware.
## Android & ONNX Runtime
 * **ONNX Runtime for Android Documentation**
   Since you are already using ONNX for your text-to-speech engine in Jandal AI, this documentation is crucial. It details how to optimize the runtime footprint, handle audio buffer streams, and manage Opset version compatibility.
 * **TensorFlow Lite (LiteRT) Audio Classification**
   If you decide to pivot back to microWakeWord (which outputs .tflite models), this documentation explains how to structure Android microphone buffers (typically 16 kHz, 16-bit PCM) for real-time local inference.
## Home Assistant Integration
 * **Home Assistant: Year of the Voice**
   The official hub for all things voice in Home Assistant. It covers the Assist pipeline architecture, how wake words interact with the system, and how to configure text-to-speech.
 * **Wyoming Satellite on GitHub**
   If you plan to use an Android device simply as a "dumb" microphone that streams audio back to your CachyOS server (rather than doing the wake word processing natively inside Jandal AI), this project shows how the Wyoming protocol handles the audio streaming and wake word remote detection.

**Comment by lokhor** [comment]
## Progress update — 2026-05-27

### What's done

**Data generation (Steps 1–2) is complete locally.**

- ✅ `training/wakeword/generate_tts_clips.py` — Piper TTS positive clip generator (multi-voice, augmented)
- ✅ `training/wakeword/generate_hey_jandal.py` — Qwen3-TTS voice-clone clip generator
- ✅ `training/wakeword/data/positives/` — **954 clips** ready (174 from real recordings × augmentation + 780 TTS-synthesised across 11 Piper voices, 13 phonetic variants, augment_factor=5)
- ✅ `training/wakeword/data/real_recordings/` — 29 real recordings of "Hey Jandal" in the owner's voice
- ✅ `training/wakeword/piper_voices/` — 11 Piper ONNX voices downloaded locally

**All of the above is untracked (not yet committed to main).** The training scripts exist on branch `feature/985-wake-word-detection-service` but data and venv are local only.

### What's missing

- ❌ `models/wakeword/melspectrogram.onnx` — not downloaded → `WakeWordService` exits immediately; Hey Jandal toggle reports unavailable
- ❌ `models/wakeword/embedding_model.onnx` — not downloaded
- ❌ `models/wakeword/hey_jandal.onnx` — not trained

### Next steps

1. **Commit training infra to a feature branch** — scripts, requirements, README; exclude `data/`, `piper_voices/` (binary blobs) via `.gitignore`
2. **Write `train.py`** — non-interactive wrapper around the openWakeWord training notebook so Step 3 can be run as a single script without Jupyter
3. **Download base ONNX models** (`melspectrogram.onnx`, `embedding_model.onnx`) from openWakeWord releases
4. **Run training** — produce `hey_jandal.onnx` from the 954 positive clips + openWakeWord negative corpus
5. **Deploy** — copy all three ONNX files to `app/src/main/assets/models/wakeword/`, install APK to S23 Ultra, verify toggle enables and device wakes on "Hey Jandal"

### Notes on positive data quality

The 29 real recordings cover a range of distances and speaking volumes. TTS clips use 13 phonetic variants (including "Jandel", "Handal", "Handel") specifically to help the model generalise across the NZ /a/ vowel. Speed perturbation ±15%, AWGN noise SNR 10–30 dB applied per clip.

Negative examples will use the openWakeWord built-in background corpus (speech + ambient). Similar-sounding phrases ("hey Daniel", "hey Sandra", "hey Janet", "hey candle") should be verified against the trained model's false positive rate before shipping — tracked in #986.
